File: src/minecraft/monnef/jaffas/xmas/block/ItemBlockXmasMulti.java
/*
 * Jaffas and more!
 * author: monnef
 */

package monnef.jaffas.xmas.block;

import cpw.mods.fml.common.registry.LanguageRegistry;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import monnef.core.common.CustomIconHelper;
import monnef.jaffas.food.block.ItemBlockJaffas;
import monnef.jaffas.xmas.common.Reference;
import net.minecraft.client.renderer.texture.IconRegister;
import net.minecraft.item.ItemStack;
import net.minecraft.util.Icon;
import net.minecraft.util.MathHelper;

public abstract class ItemBlockXmasMulti extends ItemBlockJaffas {
    private String[] subTitles;

    public ItemBlockXmasMulti(int id) {
        super(id);
        setHasSubtypes(true);
        this.subNames = this.getSubNames();
        this.subTitles = this.getSubTitles();
        this.registerNames(getParentBlock());
    }

    @Override
    public int getMetadata(int damageValue) {
        return damageValue;
    }

    public int getSubBlocksCount() {
        return this.subNames.length;
    }

    public void registerNames(BlockXmasMulti block) {
        for (int i = 0; i < subNames.length; i++) {
            ItemStack multiBlockStack = new ItemStack(block, 1, i);
            LanguageRegistry.addName(multiBlockStack, subTitles[multiBlockStack.getItemDamage()]);
        }
    }

    @SideOnly(Side.CLIENT)
    @Override
    public Icon getIconFromDamage(int par1) {
        int var2 = MathHelper.clamp_int(par1, 0, subNames.length);
        return icons[var2];
    }

    @Override
    public String getDefaultModName() {
        return Reference.ModName;
    }

    @Override
    public int getDefaultSheetNumber() {
        return 4;
    }

    public abstract String[] getSubNames();

    public abstract String[] getSubTitles();

    public abstract BlockXmasMulti getParentBlock();

    public Icon[] icons;

    @Override
    public void registerIcons(IconRegister register) {
        icons = new Icon[getSubBlocksCount()];
        for (int i = 0; i < icons.length; i++) {
            icons[i] = register.registerIcon(CustomIconHelper.generateShiftedId(this, i));
        }
    }
}

File: src/minecraft/monnef/jaffas/xmas/block/ItemBlockXmasMulti.scala
package monnef.jaffas.xmas.block

import monnef.jaffas.food.block.ItemBlockJaffas
import net.minecraft.item.ItemStack
import cpw.mods.fml.common.registry.LanguageRegistry
import cpw.mods.fml.relauncher.{SideOnly, Side}
import net.minecraft.util.{MathHelper, Icon}
import scala.Predef.String
import monnef.core.common.CustomIconHelper
import monnef.core.api.ICustomIcon
import net.minecraft.client.renderer.texture.IconRegister
import monnef.jaffas.xmas.common.IconDescriptorXmas

abstract class ItemBlockXmasMulti(_id: Int) extends ItemBlockJaffas(_id) with IconDescriptorXmas {
  private var subTitles: Array[String] = null

  setHasSubtypes(true)
  this.subNames = this.getSubNames
  this.subTitles = this.getSubTitles
  this.registerNames(getParentBlock)

  override def getMetadata(damageValue: Int): Int = damageValue

  def getSubBlocksCount: Int = this.subNames.length

  def registerNames(block: BlockXmasMulti) {
    {
      var i: Int = 0
      while (i < subNames.length) {
        {
          val multiBlockStack: ItemStack = new ItemStack(block, 1, i)
          LanguageRegistry.addName(multiBlockStack, subTitles(multiBlockStack.getItemDamage))
        }
        i += 1
      }
    }
  }

  @SideOnly(Side.CLIENT) override def getIconFromDamage(dmg: Int): Icon = {
    val idx: Int = MathHelper.clamp_int(dmg, 0, subNames.length)
    icons(idx)
  }

  def getSubNames: Array[String]

  def getSubTitles: Array[String]

  def getParentBlock: BlockXmasMulti

  // TODO: refactor to use parent's stuff?
  //override var icons: Array[Icon] = null

  override def registerIcons(register: IconRegister) {
    icons = new Array[Icon](getSubBlocksCount)
    var i: Int = 0
    while (i < icons.length) {
      icons(i) = register.registerIcon(CustomIconHelper.generateShiftedId(this.asInstanceOf[ICustomIcon], i))
      i += 1
    }
  }
}
